delay import fixes

jeremyd2019 opened the pull request.

lhmouse committed:
bfd,ld,dlltool: Emit delay-load import data into its own section

A delay-import symbol (of a function) is resolved when a call to it is made.
The delay loader may overwrite the `__imp_` pointer to the actual function
after it has been resolved, which requires the pointer itself be in a
writeable section.

Previously it was placed in the ordinary Import Address Table (IAT), which
is emitted into the `.idata` section, which had been changed to read-only
in db00f6c3aceabbf03acdb69e74b59b2d2b043cd7, which caused segmentation
faults when functions from delay-import library were called.  This is
PR 32675.

This commit makes DLLTOOL emit delay-import IAT into `.didat`, as specified
by Microsoft. Most of the code is copied from `.idata`, except that this
section is writeable.  As a side-effect of this, PR 14339 is also fixed.

Using this DEF:

   ```
   ; ws2_32.def
   LIBRARY "WS2_32.DLL"
   EXPORTS
     WSAGetLastError
   ```

and this C program:

   ```
   // delay.c
   #define WIN32_LEAN_AND_MEAN 1
   #include <windows.h>
   #include <stdio.h>

   /////////////////////////////////////////////////////////
   // User code
   /////////////////////////////////////////////////////////

   DWORD WINAPI WSAGetLastError(void);
   extern PVOID __imp_WSAGetLastError;

   int
   main(void)
     {
       fprintf(stderr, "before delay load, __imp_WSAGetLastError = %p\n", __imp_WSAGetLastError);
       SetLastError(123);
       fprintf(stderr, "WSAGetLastError() = %d\n", WSAGetLastError());
       fprintf(stderr, "after delay load, __imp_WSAGetLastError = %p\n", __imp_WSAGetLastError);
       __imp_WSAGetLastError = (PVOID) 1234567;
       fprintf(stderr, "after plain write, __imp_WSAGetLastError = %p\n", __imp_WSAGetLastError);
     }

   /////////////////////////////////////////////////////////
   // Overridden `__delayLoadHelper2` facility
   /////////////////////////////////////////////////////////

   extern char __ImageBase[];
   PVOID WINAPI ResolveDelayLoadedAPI(PVOID ParentModuleBase, LPCVOID DelayloadDescriptor,
                                      PVOID FailureDllHook, PVOID FailureSystemHook,
                                      FARPROC* ThunkAddress, ULONG Flags);
   FARPROC WINAPI DelayLoadFailureHook(LPCSTR name, LPCSTR function);

   FARPROC WINAPI __delayLoadHelper2(LPCVOID pidd, FARPROC* ppfnIATEntry)
   {
     return ResolveDelayLoadedAPI(&__ImageBase, pidd, NULL, (PVOID) DelayLoadFailureHook,
                                  ppfnIATEntry, 0);
   }
   ```

This program used to crash:

   ```
   $ dlltool -nn -d ws2_32.def -y delay_ws2_32.a
   $ gcc -g delay.c delay_ws2_32.a -o delay.exe
   $ ./delay.exe
   before delay load, __imp_WSAGetLastError = 00007FF6937215C6
   Segmentation fault
   ```

After this commit, it loads and calls `WSAGetLastError()` properly, and
`__imp_WSAGetLastError` is writeable:

   ```
   $ dlltool -nn -d ws2_32.def -y delay_ws2_32.a
   $ gcc -g delay.c delay_ws2_32.a -o delay.exe
   $ ./delay.exe
   before delay load, __imp_WSAGetLastError = 00007FF76E2215C6
   WSAGetLastError() = 123
   after delay load, __imp_WSAGetLastError = 00007FFF191FA720
   after plain write, __imp_WSAGetLastError = 000000000012D687
   ```

Reference: https://learn.microsoft.com/en-us/windows/win32/secbp/pe-metadata#import-handling
Co-authored-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>
Signed-off-by: LIU Hao <lh_mouse@126.com>
Signed-off-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>

jeremyd2019 committed:
dlltool: respect use-nul-prefixed-import-tables option for delaylib

Noticed the extra zeros while inspecting the output.

Signed-off-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>

jeremyd2019 committed:
ld,dlltool: move read-only delayimp data into .rdata

This allows the delay IAT to be in its own section with nothing else, as
required by IMAGE_GUARD_DELAYLOAD_IAT_IN_ITS_OWN_SECTION, documented at
https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#load-configuration-layout

Signed-off-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>

jeremyd2019 committed:
bfd: populate delay import directory in PE header.

Previously, the delay import table was constructed but its rva and size
were never put into the PE optional header.

Signed-off-by: Jeremy Drake <sourceware-bugzilla@jdrake.com>

lhmouse commented:
_Moving previous discussion here:_

I kinda agree that ideally the read-only parts, such as DLL names, Import Lookup Tables, and Hint/Name Table, should be combined to `.rdata`, like MSVC does.

My previous patch didn't merge `.idata` into `.rdata`. When I attempted to do that, the program crashed upon startup, which I didn't look into it further. In ld/scripttempl/pe.sc, I made the structure of `.didat` similar to `.idata` to reduce inconsistency.

jeremyd2019 commented:
Setting the section names like in the wine MR https://gitlab.winehq.org/wine/wine/-/merge_requests/7328/diffs, pro is that it is more clear where they're ending up, but con is that I had to put them at the start of `.rdata` because otherwise the wildcard `.rdata$*` picked them up and messed up the order, resulted in the start/end markers not working so the delay directory was not set and there were not the terminating 0s of the didat2 section.

I think I'm going back to the `.didat$n` names, that seemed cleaner.

lhmouse commented on ld/scripttempl/pep.sc:
```diff
@@ -41,10 +41,12 @@ if test "${RELOCATING}"; then
+    __DELAY_IMPORT_DIRECTORY_end__ = .;
     /* These zeroes mark the end of the import list.  */
-    LONG (0); LONG (0); LONG (0); LONG (0); LONG (0);
+    . += (__DELAY_IMPORT_DIRECTORY_end__ - __DELAY_IMPORT_DIRECTORY_start__) ? 8*4 : 0;
```
[Microsoft documentation](https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#the-idata-section) says the Directory Table (`.idata$2`) is terminated by a null Directory Entry, which is 20 bytes (`sizeof(IMAGE_IMPORT_DESCRIPTOR)`) for both PE and PE+, but I think you can just hard-code `20` as it can't change.

jeremyd2019 replied:
yes, that would explain the 5 longs that were copied from the `.idata` section.  But this is [the Delay-Load Directory Table](https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#the-delay-load-directory-table), which is 32 bytes (8 4-byte entries, where the 8*4 came from), and I wanted to avoid filling in a bunch of zeros if there were no delay import directory entries so I did that `+=` trick so I could use a ternary to avoid adding space if there is nothing between the start and end symbols.

lhmouse replied:
Ahh I see. Makes sense. Does filling zeroes require `FILL(0)`? The `.idata` section apparently does not have a default fill pattern, and I am uncertain whether the gap is filled with zeroes.

jeremyd2019 replied:
I was concerned about that too, but I didn't see any way to restore the `FILL` to its original state after setting `FILL(0)`, and when I tested on i686 at least `.rdata` was filling with zeros already.

jeremyd2019 commented:
@lhmouse do you want to make any revisions to your patch on this branch?  The only thing I'm thinking is to add references to the commit message that it fixes https://sourceware.org/bugzilla/show_bug.cgi?id=14339 and https://sourceware.org/bugzilla/show_bug.cgi?id=32675

lhmouse commented:
I think we had better squash these patches. You may add this header:

* `Co-authored-by: LIU Hao <lh_mouse@126.com>`

jeremyd2019 commented:
I think I'd send them separate first and ask if they want them squashed.  It'd be easier to squash them than to try to split things out again later, especially with one patch by a different author.

jeremyd2019 commented:
BTW, I'm waiting to see the patches from #2 pushed so I can rebase this on top before sending

lhmouse commented:
> I think I'd send them separate first and ask if they want them squashed. It'd be easier to squash them than to try to split things out again later, especially with one patch by a different author.

I think there were mistakes in my first patch, which were likely fixed by yours, so I kinda hope they are squashed; there's no reason to keep an incorrect patch in commit history.

Using `Co-authored-by:` is acceptable, and GitHub recognizes it too.

jeremyd2019 commented:
https://inbox.sourceware.org/binutils/e2ac90bc-0fba-ccd0-c7a1-63c37fa7597c@huh/T/#u

jeremyd2019 commented:
I've addressed the comments on my two patches.  @lhmouse would you address https://inbox.sourceware.org/binutils/b49a89fc-b1aa-43ce-ad80-4c3a90d683f0@suse.com/T/#t ?  I also think you should claim the fix for https://sourceware.org/bugzilla/show_bug.cgi?id=14339 and https://sourceware.org/bugzilla/show_bug.cgi?id=32675 in your commit message since your commit would fix those.

lhmouse commented:
Do you mean the `startswith()` thing? I think we can add a new function for that purpose:

```c
static inline bool
is_subsection (const char *str, const char *prefix)
{
  size_t n = strlen (prefix);
  return strncmp (str, prefix, n) == 0
         && (str[n] == 0 || str[n] == '$');
}
```

You can edit my patch as you see fit. Actually I wasn't aware of https://sourceware.org/bugzilla/show_bug.cgi?id=14339 .

jeremyd2019 commented:
This is merged upstream now

jeremyd2019 commented:
What's the procedure for the bugzilla bugs?  Am I supposed to resolve them or is that done at release?

jeremyd2019 commented:
logged back in, and turns out I can't resolve them, so I guess somebody else will do that...

lhmouse commented:
> What's the procedure for the bugzilla bugs?  Am I supposed to resolve them or is that done at release?

I suggest you ask the author of that bug to check whether the issue has been resolved. The procedure varies; some people who have write access to both Git and bugzilla may close bugs right after they commit a fix, but others don't.
